pkgs/dart_model/lib/src/json_buffer.dart

```diff
@@ -0,0 +1,390 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:typed_data';
+
+/// Map with lazily evaluated values.
+///
+/// This is the efficient way to accumulate data into a [JsonBuffer]. Nested
+/// maps should also be `LazyMap`.
+class LazyMap with MapMixin<String, Object?> implements Map<String, Object?> {
+  @override
+  final List<String> keys;
+  final Object? Function(String) lookup;
+
+  LazyMap(Iterable<String> keys, this.lookup)
+      : keys = UnmodifiableListView(keys.toList());
+
+  @override
+  Object? operator [](Object? key) => key is String ? lookup(key) : null;
```

Comment:
consider throwing instead of returning null, this is really an error

Reply:
The `Map` interface disagrees, but this is a specialist enough use case that we might as well try being strict, done. (I've never really liked `Object?` here although there are definitely arguments for it).